pkg/sentry/socket/plugin/stack/socket.go

```diff
@@ -0,0 +1,886 @@
+	bytes := dst.NumBytes()
+	if bytes <= 0 {
+		iovs := iovecsFromBlockSeq(safemem.BlockSeq{}, nil)
+		rc, la, lc, flag := cgo.Recvmsg(s.fd, iovs, addr, control, sysflags)
+		addr = addr[:la]
+		control = control[:lc]
+		ret, err := translateReturn(rc)
+		return int64(ret), err, addr, control, flag
+	}
+
+	// Slow path for non-stream socket(e.g., UDP, raw socket).
+	if s.skType != linux.SOCK_STREAM {
+		if len(addr) == 0 && s.skType == linux.SOCK_DGRAM {
+			addr = make([]byte, sizeofSockaddr)
+		}
+
+		tmpBuf := make([]byte, bytes)
+		tmpBS := safemem.BlockSeqOf(safemem.BlockFromSafeSlice(tmpBuf))
+		iovs := iovecsFromBlockSeq(tmpBS, nil)
+		rc, la, lc, flag := cgo.Recvmsg(s.fd, iovs, addr, control, sysflags)
+		if rc < 0 {
+			_, err := translateReturn(rc)
+			return 0, err, addr /* return the original slice */, control /* return the original slice */, 0
+		}
+
+		n, err := dst.CopyOut(t, tmpBuf[:rc])
+		return int64(n), err, addr[:la], control[:lc], flag
+	}
+
+	// Fast path for stream socket(e.g., TCP socket).
+	rw := getReadWriter(s.fd)
+	rw.to = control
+	rw.flags = uint32(sysflags)
+
+	n, err := dst.CopyOutFrom(t, rw)
+
+	control = rw.to
+	rw.to = nil
+	msg_flags := int(rw.flags)
+	rw.flags = 0
+	putReadWriter(rw)
+
+	return n, err, nil /* ignore for TCP */, control, msg_flags
+}
+
+// RecvMsg implements socket.Socket.RecvMsg.
+func (s *socketOperations) RecvMsg(t *kernel.Task, dst usermem.IOSequence, flags int, haveDeadline bool, deadline ktime.Time, senderRequested bool, controlDataLen uint64) (int, int, linux.SockAddr, uint32, socket.ControlMessages, *syserr.Error) {
+	var (
+		senderAddr linux.SockAddr
+		addr       []byte
+		addrlen    uint32
+		control    []byte
+		nonblock   bool
+		n          int64
+		err        error
+		mflag      int
+	)
+	if senderRequested {
+		if s.skType != linux.SOCK_STREAM {
+			addr = make([]byte, sizeofSockaddr)
+		} else {
+			// According to UNIX98, msg_name/msg_namelen are ignored on connected socket.
+			senderRequested = false
+		}
+	}
+	if controlDataLen > 0 {
+		control = make([]byte, controlDataLen)
+	}
+
+	if (flags & syscall.MSG_DONTWAIT) != 0 {
+		nonblock = true
+	}
+	waitall := (nonblock == false) && (flags&syscall.MSG_WAITALL != 0)
+	sysflags := flags | syscall.MSG_DONTWAIT // always non-blocking
+
+	n, err, addr, control, mflag = s.recv(t, dst, sysflags, addr, control)
+
+	if nonblock {
+		if err != nil {
+			// avoid invalid controlMessages when error occurs
+			control = nil
+		} else {
+			if senderRequested && len(addr) > 0 {
+				senderAddr = socket.UnmarshalSockAddr(s.family, addr)
+				addrlen = uint32(len(addr))
+			}
+		}
+		controlMessages := buildControlMessage(control)
+		if err == error(syscall.ESHUTDOWN) {
+			err = linuxerr.ErrWouldBlock
+		}
+		return int(n), mflag, senderAddr, addrlen, *controlMessages, syserr.FromError(err)
+	}
+
+	rn := n
+	for err == linuxerr.ErrWouldBlock || (waitall && err == nil && rn < dst.NumBytes()) {
```

Comment:
This for loop is doing blocking RecvMsg. It will try to read all data until either all data have been received or an error is reported.
This ErrWouldBlock error will be generated when we fail to get a POLLIN event reported before timeout.